src/mds/journal.cc

```diff
@@ -237,27 +237,25 @@ void LogSegment::try_to_expire(MDSRank *mds, MDSGatherBuilder &gather_bld, int o
 
   ceph_assert(g_conf()->mds_kill_journal_expire_at != 3);
 
-  size_t count = 0;
-  for (elist<CInode*>::iterator it = dirty_parent_inodes.begin(); !it.end(); ++it)
-    count++;
-
-  std::vector<CInodeCommitOperations> ops_vec;
-  ops_vec.reserve(count);
+  std::map<int64_t, std::vector<CInodeCommitOperations>> ops_vec_map;
   // backtraces to be stored/updated
   for (elist<CInode*>::iterator p = dirty_parent_inodes.begin(); !p.end(); ++p) {
     CInode *in = *p;
     ceph_assert(in->is_auth());
     if (in->can_auth_pin()) {
       dout(15) << "try_to_expire waiting for storing backtrace on " << *in << dendl;
-      ops_vec.resize(ops_vec.size() + 1);
-      in->store_backtrace(ops_vec.back(), op_prio);
+      auto pool_id = in->get_backtrace_pool();
```

Comment:
I realized that this change doesn't fully fix the issue - CInode::_store_backtrace() will iterate the old_pools list and queue a backtrace update on those, so, if the main backtrace pool doesn't exist (deleted data pool), then file system will do read-only.

Fix would be to queue separate backtrace update operation for old pools.

Reply:
> I realized that this change doesn't fully fix the issue - CInode::_store_backtrace() will iterate the old_pools list and queue a backtrace update on those, so, if the main backtrace pool doesn't exist (deleted data pool), then file system will do read-only.

"main backtrace pool" == "default data pool"?

Reply:
I meant if one of those old pools were non existent then the problem of MDS going read-only will surface again. So, the fix is to dispatch separate backtrace updated for each pool.